File: 01-week-1--starter-algorithms/02-day-3--fibonacci-series/javascript/fibonacci.js
function fibonacci(num) {
  if (num < 2){
    return num
    // the edge case if the value is less than 2
    // prevents us from running unecessary code/iterations
  } else {
    let inputAry= [0,1]
    // starts out the fibonacci sequence, it is known
    for(let i=1; i < num; i++){
      let fibForm = inputAry[i] + inputAry[i-1]
      // formula to get the numbers from the fibonacci sequence. we add the 2 preceeding numbers in the array
      inputAry.push(fibForm)
      // we add the new fib number onto the array for storage
    } 
    return inputAry[num]
  }
}

//   /(\
//   ¡ !´\
//   | )\ `.
//   | `.)  \,-,--
//   (      / /
//    `'-.,;_/
//           `----

// function fibonacci(num) {
//   if (num < 2){
//     return num
//   } else {
//       let inputAry= [0,1]
//       for(let i= 0; i < num-1 ; i++){
//         let theSum = inputAry[0] + inputAry[1]
//         inputAry= [inputAry[1], theSum]
//       }
//       return inputAry[1]
//   }
// }

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log("=>", fibonacci(0));

  console.log("");

  console.log("Expecting: 1");
  console.log("=>", fibonacci(2));

  console.log("");

  console.log("Expecting: 55");
  console.log("=>", fibonacci(10));
}

module.exports = fibonacci;

// Please add your pseudocode to this file
// And a written explanation of your solution
// return an output in the fibonacci sequence based off the indexed number (input)
// the third number in a fib seq, is the sum of the 2 preceeding numbers
// create an array to hold the values from input
// create a formula that adds new values into the fib sequence and adds it to the array variable
// iterate that formula by the times from the input
// grab the number at the given index from the created array

File: 01-week-1--starter-algorithms/02-day-3--fibonacci-series/ruby/fibonacci.rb
def fibonacci(num)
  if num < 2
    return num
    #edge case to avoid expensive iteration if we are given num < 2
  else
    input_ary= [0,1]
    # create a starting point for fib seq
    (num-1).times do
      # it iterates one number fewer, instead of going 1 num over if you did
      ## num.times do, since we already created the first sum, we need to iterate one less
      sum = input_ary[0]+ input_ary[1]
      # creating a new based on the 2 preceeding nums in the array to add to our input_ary variable
      input_ary = [input_ary[1], sum]
      # destructively updates/re-defines the input_array
      
    end
    return input_ary.last
    # returns the second value in our input array-- since our input_array only 
    # has two vals- preceeding num and sum. our input array only stores two values to avoid a bunch of numbers we dont need
  end

  # type your code in here
end


# /(\
# ||\\
# ! ‾ `;-
#  `--/__
# get rid of the edge case- where the input is less than 2. return the first vals in fib seq.
# create a temp ary to store values
#initialize the temp ary with firsttwo vals in fib sequence
#iterate through the ary num times
# return the num at the index

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts "=>", fibonacci(0)

  puts

  puts "Expecting: 1"
  puts "=>", fibonacci(2)

  puts

  puts "Expecting: 55"
  puts "=>", fibonacci(10)

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
